1. LOGISTIC REGRESSION

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df=sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [ ]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [ ]:
df.drop(['deck','embark_town', 'parch', 'who', 'embark_town'],axis=1,inplace=True)

In [ ]:
df

survived  pclass     sex   age  sibsp     fare embarked   class  \
0           0       3    male  22.0      1   7.2500        S   Third   
1           1       1  female  38.0      1  71.2833        C   First   
2           1       3  female  26.0      0   7.9250        S   Third   
3           1       1  female  35.0      1  53.1000        S   First   
4           0       3    male  35.0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...      ...      ...     ...   
886         0       2    male  27.0      0  13.0000        S  Second   
887         1       1  female  19.0      0  30.0000        S   First   
888         0       3  female   NaN      1  23.4500        S   Third   
889         1       1    male  26.0      0  30.0000        C   First   
890         0       3    male  32.0      0   7.7500        Q   Third   

     adult_male alive  alone  
0          True    no  False  
1         False   yes  False  
2         False   yes   True  
3         False   yes  False  
4          True    no   True  
..          ...   ...    ...  
886        True    no   True  
887       False   yes   True  
888       False    no  False  
889        True   yes   True  
890        True    no   True  

[891 rows x 11 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   survived    891 non-null    int64   
 1   pclass      891 non-null    int64   
 2   sex         891 non-null    object  
 3   age         714 non-null    float64 
 4   sibsp       891 non-null    int64   
 5   fare        891 non-null    float64 
 6   embarked    889 non-null    object  
 7   class       891 non-null    category
 8   adult_male  891 non-null    bool    
 9   alive       891 non-null    object  
 10  alone       891 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(3), object(3)
memory usage: 58.6+ KB


In [ ]:
df['age'].fillna(df['age'].mean(),inplace=True)

In [ ]:
df.dropna(subset='embarked',inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   survived    889 non-null    int64   
 1   pclass      889 non-null    int64   
 2   sex         889 non-null    object  
 3   age         889 non-null    float64 
 4   sibsp       889 non-null    int64   
 5   fare        889 non-null    float64 
 6   embarked    889 non-null    object  
 7   class       889 non-null    category
 8   adult_male  889 non-null    bool    
 9   alive       889 non-null    object  
 10  alone       889 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(3), object(3)
memory usage: 65.2+ KB


In [ ]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [ ]:
df

survived  pclass     sex        age  sibsp     fare embarked   class  \
0           0       3    male  22.000000      1   7.2500        S   Third   
1           1       1  female  38.000000      1  71.2833        C   First   
2           1       3  female  26.000000      0   7.9250        S   Third   
3           1       1  female  35.000000      1  53.1000        S   First   
4           0       3    male  35.000000      0   8.0500        S   Third   
..        ...     ...     ...        ...    ...      ...      ...     ...   
886         0       2    male  27.000000      0  13.0000        S  Second   
887         1       1  female  19.000000      0  30.0000        S   First   
888         0       3  female  29.699118      1  23.4500        S   Third   
889         1       1    male  26.000000      0  30.0000        C   First   
890         0       3    male  32.000000      0   7.7500        Q   Third   

     adult_male alive  alone  
0          True    no  False  
1         False   yes  False  
2         False   yes   True  
3         False   yes  False  
4          True    no   True  
..          ...   ...    ...  
886        True    no   True  
887       False   yes   True  
888       False    no  False  
889        True   yes   True  
890        True    no   True  

[889 rows x 11 columns]

In [ ]:
df['sex']=le.fit_transform(df['sex'])

In [ ]:
df['embarked']=le.fit_transform(df['embarked'])
df['class']=le.fit_transform(df['class'])
df['adult_male']=le.fit_transform(df['adult_male'])
df['alive']=le.fit_transform(df['alive'])
df['alone']=le.fit_transform(df['alone'])

In [ ]:
df

survived  pclass  sex        age  sibsp     fare  embarked  class  \
0           0       3    1  22.000000      1   7.2500         2      2   
1           1       1    0  38.000000      1  71.2833         0      0   
2           1       3    0  26.000000      0   7.9250         2      2   
3           1       1    0  35.000000      1  53.1000         2      0   
4           0       3    1  35.000000      0   8.0500         2      2   
..        ...     ...  ...        ...    ...      ...       ...    ...   
886         0       2    1  27.000000      0  13.0000         2      1   
887         1       1    0  19.000000      0  30.0000         2      0   
888         0       3    0  29.699118      1  23.4500         2      2   
889         1       1    1  26.000000      0  30.0000         0      0   
890         0       3    1  32.000000      0   7.7500         1      2   

     adult_male  alive  alone  
0             1      0      0  
1             0      1      0  
2             0      1      1  
3             0      1      0  
4             1      0      1  
..          ...    ...    ...  
886           1      0      1  
887           0      1      1  
888           0      0      0  
889           1      1      1  
890           1      0      1  

[889 rows x 11 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [ ]:
df['age']=sc.fit_transform(df[['age']])
df['fare']=sc.fit_transform(df[['fare']])
df['pclass']=sc.fit_transform(df[['pclass']])

In [ ]:
df

survived    pclass  sex       age  sibsp      fare  embarked  class  \
0           0  0.825209    1 -0.590495      1 -0.500240         2      2   
1           1 -1.572211    0  0.643971      1  0.788947         0      0   
2           1  0.825209    0 -0.281878      0 -0.486650         2      2   
3           1 -1.572211    0  0.412509      1  0.422861         2      0   
4           0  0.825209    1  0.412509      0 -0.484133         2      2   
..        ...       ...  ...       ...    ...       ...       ...    ...   
886         0 -0.373501    1 -0.204724      0 -0.384475         2      1   
887         1 -1.572211    0 -0.821957      0 -0.042213         2      0   
888         0  0.825209    0  0.003524      1 -0.174084         2      2   
889         1 -1.572211    1 -0.281878      0 -0.042213         0      0   
890         0  0.825209    1  0.181046      0 -0.490173         1      2   

     adult_male  alive  alone  
0             1      0      0  
1             0      1      0  
2             0      1      1  
3             0      1      0  
4             1      0      1  
..          ...    ...    ...  
886           1      0      1  
887           0      1      1  
888           0      0      0  
889           1      1      1  
890           1      0      1  

[889 rows x 11 columns]

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x=df.drop('survived',axis=1)
y=df['survived']

In [ ]:
x.drop(['alive', 'alone', 'adult_male'], axis=1, inplace=True)

In [ ]:
x

pclass  sex       age  sibsp      fare  embarked  class
0    0.825209    1 -0.590495      1 -0.500240         2      2
1   -1.572211    0  0.643971      1  0.788947         0      0
2    0.825209    0 -0.281878      0 -0.486650         2      2
3   -1.572211    0  0.412509      1  0.422861         2      0
4    0.825209    1  0.412509      0 -0.484133         2      2
..        ...  ...       ...    ...       ...       ...    ...
886 -0.373501    1 -0.204724      0 -0.384475         2      1
887 -1.572211    0 -0.821957      0 -0.042213         2      0
888  0.825209    0  0.003524      1 -0.174084         2      2
889 -1.572211    1 -0.281878      0 -0.042213         0      0
890  0.825209    1  0.181046      0 -0.490173         1      2

[889 rows x 7 columns]

In [ ]:
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 889, dtype: int64

In [ ]:
x_train, x_test, y_train, y_test=train_test_split(x,y, test_size=0.2, random_state=42)

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
model=LogisticRegression()
model.fit(x_train, y_train)

LogisticRegression()

In [ ]:
y_pred=model.predict(x_test)
y_pred

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
y_test

281    0
435    1
39     1
418    0
585    1
      ..
433    0
807    0
25     1
85     1
10     1
Name: survived, Length: 178, dtype: int64

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
accuracy_score(y_pred, y_test)

0.7752808988764045

In [ ]:
confusion_matrix(y_test, y_pred, labels=[1,0])

array([[54, 15],
       [25, 84]])

In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.77      0.81       109
           1       0.68      0.78      0.73        69

    accuracy                           0.78       178
   macro avg       0.77      0.78      0.77       178
weighted avg       0.78      0.78      0.78       178



2. kNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
knn_model=KNeighborsClassifier(n_neighbors=5)
knn_model.fit(x_train, y_train) #we did not scaled y train because it has output variablle 'survived' which is already in 0 & 1 form

KNeighborsClassifier()

In [ ]:
y_pred_knn=knn_model.predict(x_test)
y_pred_knn

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
y_test

281    0
435    1
39     1
418    0
585    1
      ..
433    0
807    0
25     1
85     1
10     1
Name: survived, Length: 178, dtype: int64

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

accuracy_score(y_test, y_pred_knn)

0.7921348314606742

In [ ]:
confusion_matrix(y_test, y_pred_knn)

array([[88, 21],
       [16, 53]])

In [ ]:
print(classification_report(y_test, y_pred_knn))

              precision    recall  f1-score   support

           0       0.85      0.81      0.83       109
           1       0.72      0.77      0.74        69

    accuracy                           0.79       178
   macro avg       0.78      0.79      0.78       178
weighted avg       0.80      0.79      0.79       178



3. NAIVE BAYES

Usually standard scaling data is not required in naive bayes model, it is required in knn, but i have already done standard scaling above , so i will use the same for naive bayes


In [ ]:
from sklearn.naive_bayes import GaussianNB

In [ ]:
model_NB = GaussianNB() # Added parentheses to create an instance
model_NB.fit(x_train, y_train)

GaussianNB()

In [ ]:
y_pred_NB=model_NB.predict(x_test)
y_pred_NB

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0,
       0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
y_test

281    0
435    1
39     1
418    0
585    1
      ..
433    0
807    0
25     1
85     1
10     1
Name: survived, Length: 178, dtype: int64

In [ ]:
accuracy_score(y_test, y_pred_NB)

0.7528089887640449

In [ ]:
confusion_matrix(y_test, y_pred_NB)

array([[81, 28],
       [16, 53]])

In [ ]:
print(classification_report(y_test, y_pred_NB))

              precision    recall  f1-score   support

           0       0.84      0.74      0.79       109
           1       0.65      0.77      0.71        69

    accuracy                           0.75       178
   macro avg       0.74      0.76      0.75       178
weighted avg       0.76      0.75      0.76       178



4. DECISION TREE

Decision tree require scaled data

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
model_decision_tree=DecisionTreeClassifier()
model_decision_tree.fit(x_train, y_train)

DecisionTreeClassifier()

In [ ]:
y_pred_decision_tree=model_decision_tree.predict(x_test)
y_pred_decision_tree

array([1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
y_test

281    0
435    1
39     1
418    0
585    1
      ..
433    0
807    0
25     1
85     1
10     1
Name: survived, Length: 178, dtype: int64

In [ ]:
accuracy_score(y_test, y_pred_decision_tree)

0.7640449438202247

In [ ]:
confusion_matrix(y_test, y_pred_decision_tree)

array([[81, 28],
       [14, 55]])

In [ ]:
print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       0.85      0.74      0.79       109
           1       0.66      0.80      0.72        69

    accuracy                           0.76       178
   macro avg       0.76      0.77      0.76       178
weighted avg       0.78      0.76      0.77       178



In decision tree model, for continous variable(feature) in data like age and fare in above dataset, the entropy and information gain is calculated through MEan Square Error(MSE), sum of residuals,


Temperature	16	18	20	22	25	30
Play?	No	No	No	Yes	Yes	Yes
🎯 Step 1: Find ALL Possible Split Points
We look between each pair of different temperatures:

Possible split points:

Between 16 & 18 → 17

Between 18 & 20 → 19 ← This is where 19 came from!

Between 20 & 22 → 21

Between 22 & 25 → 23.5

Between 25 & 30 → 27.5

We try EVERY one of these!

🔍 Step 2: Let's Try Split at 19 (Your Question!)
Rule: Temperature < 19 vs Temperature ≥ 19

Left Side (Temp < 19):
Temperatures: 16, 18

Play?: No, No

All "No" → Perfectly pure! 🎯

Entropy = 0

Right Side (Temp ≥ 19):
Temperatures: 20, 22, 25, 30

Play?: No, Yes, Yes, Yes

3 "Yes", 1 "No" → Mixed! 😕

Entropy = 0.811

Weighted Average:
$= \frac{2}{6} \times 0 + \frac{4}{6} \times 0.811 = 0.541$

Information Gain = 1.0 - 0.541 = 0.459

🏆 Step 3: Now Try Split at 21 (The Winner!)
Rule: Temperature < 21 vs Temperature ≥ 21

Left Side (Temp < 21):
Temperatures: 16, 18, 20

Play?: No, No, No

All "No" → Perfect! 🎯

Entropy = 0

Right Side (Temp ≥ 21):
Temperatures: 22, 25, 30

Play?: Yes, Yes, Yes

All "Yes" → Perfect! 🎯

Entropy = 0

Weighted Average:
$= \frac{3}{6} \times 0 + \frac{3}{6} \times 0 = 0$

Information Gain = 1.0 - 0 = 1.0 ← PERFECT! 🏆

🤔 Why Do We Try All These Points?
Because we're looking for the magic temperature that best separates "Yes" from "No"!

<19: Not good enough (still have mixed group)

<21: PERFECT! (completely separates Yes/No)

<24: Not good enough

🧠 Simple Rule:
For continuous data, we test EVERY possible "cutoff" between different values to find the one that creates the purest groups!

We tried: 17, 19, 21, 23.5, 27.5
Winner: 21 because it perfectly separates all "No" (left) from all "Yes" (right)

🎯 Visual Summary:
text
Temperatures: 16(N) 18(N) 20(N) | 22(Y) 25(Y) 30(Y)
                 All "No"        |      All "Yes"
               ← Split at 21 →   
Left of 21 = All "No"
Right of 21 = All "Yes"

That's why 21 is the perfect split point! 🎉

Pre Pruning in Decision Tree

In [ ]:
#Preprununing:
parameter= {
    'criterion': ['gini', 'entrophy', 'log_loss'],
    'splitter' :['best', 'random'],
    'max_depth' : [1,2,3,4,5],
    'max_features': ['auto', 'sqrt']
}

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
model_prune= DecisionTreeClassifier()
pruning=GridSearchCV(model_prune, parameter, cv=5, scoring='accuracy')

In [ ]:
pruning.fit(x_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entrophy', 'log_loss'],
                         'max_depth': [1, 2, 3, 4, 5],
                         'max_features': ['auto', 'sqrt'],
                         'splitter': ['best', 'random']},
             scoring='accuracy')

In [ ]:
pruning.best_params_

{'criterion': 'log_loss',
 'max_depth': 5,
 'max_features': 'sqrt',
 'splitter': 'best'}

In [ ]:
y_DT_prune=pruning.predict(x_test)
y_DT_prune

array([0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       0, 0])

In [ ]:
pruning.cv_results_

{'mean_fit_time': array([0.00192733, 0.00183668, 0.0045465 , 0.00414996, 0.00140123,
        0.00135818, 0.00474315, 0.00575528, 0.00141726, 0.00133643,
        0.00413251, 0.00393891, 0.00138597, 0.00137506, 0.00347309,
        0.0024096 , 0.00081382, 0.00086884, 0.00258703, 0.00238028,
        0.00084376, 0.00083909, 0.00079455, 0.00081139, 0.00080614,
        0.0008287 , 0.00082479, 0.00078921, 0.00083499, 0.00079923,
        0.00079322, 0.00080743, 0.00083408, 0.00086694, 0.00081115,
        0.00081511, 0.00079436, 0.00082741, 0.00080042, 0.00081754,
        0.0008265 , 0.00088363, 0.00244412, 0.00237441, 0.00084496,
        0.000842  , 0.00247641, 0.00242019, 0.00098238, 0.00084443,
        0.00263901, 0.00238957, 0.00083714, 0.00085955, 0.00274229,
        0.00244884, 0.00090523, 0.00083466, 0.00273137, 0.00254345]),
 'std_fit_time': array([2.93544471e-04, 2.39570913e-04, 1.65244998e-04, 6.77765608e-05,
        7.04135923e-05, 7.13291013e-05, 9.89599828e-04, 3.51805490e-03,
     

In [ ]:
accuracy_score(y_test, y_DT_prune)

0.7921348314606742

5.SUPPORT VECTOR MACHINES(SVM)

SVM also requires scaled data

In [ ]:
from sklearn.svm import SVC #support vector classification

In [ ]:
model_svm=SVC(kernel='rbf')
model_svm.fit(x_train, y_train)

SVC()

In [ ]:
y_pred_svm=model_svm.predict(x_test)
y_pred_svm

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
y_test

281    0
435    1
39     1
418    0
585    1
      ..
433    0
807    0
25     1
85     1
10     1
Name: survived, Length: 178, dtype: int64

In [ ]:
accuracy_score(y_test, y_pred_svm)

0.8202247191011236

In [ ]:
confusion_matrix(y_test, y_pred_svm)

array([[94, 15],
       [17, 52]])

In [ ]:
print(classification_report(y_test,y_pred_svm))

              precision    recall  f1-score   support

           0       0.85      0.86      0.85       109
           1       0.78      0.75      0.76        69

    accuracy                           0.82       178
   macro avg       0.81      0.81      0.81       178
weighted avg       0.82      0.82      0.82       178



SVM has highest accuracy 82 % among all classification models

ML 4 -- CROSS VALIDATION for SVM model

In [ ]:
X=df.drop('survived', axis=1)
X.drop(['alive', 'alone', 'adult_male'], axis=1, inplace=True)
Y=df['survived']

In [ ]:
X

pclass  sex       age  sibsp      fare  embarked  class
0    0.825209    1 -0.590495      1 -0.500240         2      2
1   -1.572211    0  0.643971      1  0.788947         0      0
2    0.825209    0 -0.281878      0 -0.486650         2      2
3   -1.572211    0  0.412509      1  0.422861         2      0
4    0.825209    1  0.412509      0 -0.484133         2      2
..        ...  ...       ...    ...       ...       ...    ...
886 -0.373501    1 -0.204724      0 -0.384475         2      1
887 -1.572211    0 -0.821957      0 -0.042213         2      0
888  0.825209    0  0.003524      1 -0.174084         2      2
889 -1.572211    1 -0.281878      0 -0.042213         0      0
890  0.825209    1  0.181046      0 -0.490173         1      2

[889 rows x 7 columns]

In [ ]:
Y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 889, dtype: int64

In [ ]:
from sklearn.model_selection import cross_val_score, cross_val_predict

In [ ]:
scaler=StandardScaler()
X_Scaled=scaler.fit_transform(X)

In [ ]:
X_Scaled

array([[ 0.82520863,  0.73534203, -0.59049493, ..., -0.50023975,
         0.58683958,  0.82520863],
       [-1.57221121, -1.35991138,  0.64397101, ...,  0.78894661,
        -1.93955453, -1.57221121],
       [ 0.82520863, -1.35991138, -0.28187844, ..., -0.48664993,
         0.58683958,  0.82520863],
       ...,
       [ 0.82520863, -1.35991138,  0.00352373, ..., -0.17408416,
         0.58683958,  0.82520863],
       [-1.57221121,  0.73534203, -0.28187844, ..., -0.0422126 ,
        -1.93955453, -1.57221121],
       [ 0.82520863,  0.73534203,  0.18104628, ..., -0.49017322,
        -0.67635748,  0.82520863]])

Prediction and score  on validation data only

In [ ]:
val_scores=cross_val_score(model_svm, x_train, y_train, cv=5, scoring='accuracy')

In [ ]:
y_val_pred=cross_val_predict(model_svm, x_train, y_train, cv=5)
y_val_pred

array([1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,

In [ ]:
y_train

708    1
240    0
382    0
792    0
683    0
      ..
107    1
271    1
862    1
436    0
103    0
Name: survived, Length: 711, dtype: int64

In [ ]:
accuracy_score(y_train, y_val_pred)

0.8087201125175809

In [ ]:
val_scores

array([0.82517483, 0.82394366, 0.81690141, 0.81690141, 0.76056338])

In [ ]:
np.mean(val_scores)

np.float64(0.808696936865951)

Prediction and score on test data only




In [ ]:
scores=cross_val_score(model_svm, X_Scaled, Y, cv=5, scoring='accuracy')  #cv is cross validation limit , like how many part I want to divide the data
#scoring will be based on accuracy
scores

#The output shows the accuracy score of different k=5 models in the dataset , as every fold was given opportunity to be part of testing data.
#And remember this all 5 accuracies are part of only one model that is SUPPORT VECTOR MACHINE (SVM)

array([0.8258427 , 0.8258427 , 0.8258427 , 0.80337079, 0.84745763])

In [ ]:
y_test_pred=model_svm.predict(x_test)
y_test_pred

array([0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1])

In [ ]:
accuracy_score(y_test, y_test_pred)

0.8202247191011236

In [ ]:
#Now we will take mean of all 5 k folds
scores.mean()

np.float64(0.8256713007046276)